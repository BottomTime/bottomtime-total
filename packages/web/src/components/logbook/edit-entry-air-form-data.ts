import { PressureUnit, TankDTO } from '@bottomtime/api';

export interface EditEntryAirFormData {
  id: string;
  tankId: string;
  tankInfo?: TankDTO;
  count: string | number;
  startPressure: string | number;
  endPressure: string | number;
  pressureUnit: PressureUnit;
  hePercent: string | number;
  o2Percent: string | number;
}
