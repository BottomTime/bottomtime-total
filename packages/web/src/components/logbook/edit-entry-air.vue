<template>
  <li class="my-1.5">
    <div class="flex justify-between">
      <TextHeading level="h4">#{{ ordinal + 1 }}</TextHeading>

      <button class="text-danger" @click="$emit('remove', ordinal)">
        <span class="sr-only">Remove air entry #{{ ordinal + 1 }}</span>
        <span>
          <i class="fa-solid fa-x"></i>
        </span>
      </button>
    </div>

    <FormBox class="grid grid-cols-2 lg:grid-cols-4 gap-2 ml-4">
      <FormField
        class="order-1 col-span-1 lg:col-span-3"
        label="Tank"
        :responsive="false"
        :invalid="v$.tankId.$error"
        :error="v$.tankId.$errors[0]?.$message"
      >
        <FormSelect
          v-model="formData.tankId"
          control-id="tanks-select"
          :options="tankOptions"
          :invalid="v$.tankId.$error"
          stretch
        />
      </FormField>

      <div
        v-if="formData.tankInfo"
        class="flex justify-evenly col-span-2 lg:col-span-4 order-3 mb-2"
      >
        <div class="text-center">
          <p class="font-bold">Working Pressure</p>
          <p class="font-mono text-sm">
            {{ formData.tankInfo.workingPressure }}bar
          </p>
        </div>
        <div class="text-center">
          <p class="font-bold">Volume</p>
          <p class="font-mono text-sm">{{ formData.tankInfo.volume }}L</p>
        </div>
        <div class="text-center">
          <p class="font-bold">Material</p>
          <p class="font-mono text-sm">
            {{ tankMaterialString }}
          </p>
        </div>
      </div>

      <FormField
        class="order-2"
        label="Count"
        :responsive="false"
        :invalid="v$.count.$error"
        :error="v$.count.$errors[0]?.$message"
      >
        <FormTextBox v-model.number="formData.count" />
      </FormField>

      <FormField
        class="order-4"
        label="Start Pressure"
        control-id="start-pressure"
        :invalid="v$.startPressure.$error"
        :error="v$.startPressure.$errors[0]?.$message"
        :responsive="false"
      >
        <div class="relative">
          <FormTextBox
            v-model.number="formData.startPressure"
            :invalid="v$.startPressure.$error"
          />
          <button
            :class="`absolute inset-y-0 end-0 rounded-r-lg border ${
              v$.startPressure.$error ? 'border-danger' : 'border-grey-950 '
            } w-10 flex justify-center items-center text-grey-950 disabled:text-grey-500 bg-secondary hover:bg-secondary-hover`"
            @click="onTogglePressureUnit"
          >
            {{ formData.pressureUnit }}
          </button>
        </div>
      </FormField>

      <FormField
        class="order-5"
        label="End Pressure"
        control-id="end-pressure"
        :responsive="false"
        :invalid="v$.endPressure.$error"
        :error="v$.endPressure.$errors[0]?.$message"
      >
        <div class="relative">
          <FormTextBox
            v-model.number="formData.endPressure"
            :invalid="v$.endPressure.$error"
          />
          <button
            :class="`absolute inset-y-0 end-0 rounded-r-lg border ${
              v$.endPressure.$error ? 'border-danger' : 'border-grey-950'
            } w-10 flex justify-center items-center text-grey-950 disabled:text-grey-500 bg-secondary hover:bg-secondary-hover`"
            @click="onTogglePressureUnit"
          >
            {{ formData.pressureUnit }}
          </button>
        </div>
      </FormField>

      <FormField
        class="order-6"
        label="O₂ %"
        control-id="o2"
        :responsive="false"
        :invalid="v$.o2Percentage.$error"
        :error="v$.o2Percentage.$errors[0]?.$message"
      >
        <div class="relative">
          <FormTextBox
            v-model.number="formData.o2Percentage"
            :invalid="v$.o2Percentage.$error"
          />
          <span
            :class="`absolute end-0 inset-y-0 font-bold text-grey-200 bg-grey-700 border dark:text-grey-200 ${
              v$.o2Percentage.$error
                ? 'border-danger'
                : 'border-grey-950 dark:bg-grey-700'
            } rounded-r-lg flex justify-center items-center w-10 pointer-events-none`"
          >
            %
          </span>
        </div>
      </FormField>

      <FormField
        class="order-7"
        label="He %"
        control-id="he"
        :responsive="false"
        :invalid="v$.hePercentage.$error"
        :error="v$.hePercentage.$errors[0]?.$message"
      >
        <div class="relative">
          <FormTextBox
            v-model.number="formData.hePercentage"
            :invalid="v$.hePercentage.$error"
          />
          <span
            :class="`absolute end-0 inset-y-0 font-bold text-grey-200 bg-grey-700 border dark:text-grey-200 ${
              v$.hePercentage.$error
                ? 'border-danger'
                : 'border-grey-950 dark:bg-grey-700'
            } rounded-r-lg flex justify-center items-center w-10 pointer-events-none`"
          >
            %
          </span>
        </div>
      </FormField>
    </FormBox>
  </li>
</template>

<script lang="ts" setup>
import { PressureUnit, TankDTO, TankMaterial } from '@bottomtime/api';

import { useVuelidate } from '@vuelidate/core';
import { between, helpers, integer, required } from '@vuelidate/validators';

import { computed, reactive, watch } from 'vue';

import { SelectOption } from '../../common';
import FormBox from '../common/form-box.vue';
import FormField from '../common/form-field.vue';
import FormSelect from '../common/form-select.vue';
import FormTextBox from '../common/form-text-box.vue';
import TextHeading from '../common/text-heading.vue';
import { EditEntryAirFormData } from './edit-entry-air-form-data';

interface EditEntryAirProps {
  air: EditEntryAirFormData;
  ordinal: number;
  tanks: TankDTO[];
}

const props = defineProps<EditEntryAirProps>();
const emit = defineEmits<{
  (e: 'remove', index: number): void;
  (e: 'update', air: EditEntryAirFormData, index: number): void;
}>();

const tankOptions = computed<SelectOption[]>(() => [
  { value: '', label: '(select tank)' },
  ...props.tanks.map((tank) => ({
    value: tank.id,
    label: tank.name,
  })),
]);

const formData = reactive<EditEntryAirFormData>({ ...props.air });

const tankMaterialString = computed(() => {
  if (!formData.tankInfo) return '';
  return formData.tankInfo.material === TankMaterial.Aluminum
    ? 'Aluminum'
    : 'Steel';
});

const v$ = useVuelidate(
  {
    tankId: {
      required: helpers.withMessage('Please select a tank', required),
    },
    count: {
      int: helpers.withMessage(
        'Number of tanks must be a whole number (e.g. 1, 2, 3, etc.)',
        integer,
      ),
      between: helpers.withMessage(
        'Number of tanks must be between 1 and 4',
        between(1, 4),
      ),
    },
    startPressure: {
      required: helpers.withMessage('Start pressure is required', required),
      valid: helpers.withMessage(
        'Start pressure must be greater than 0 and less than 300bar / 4400psi',
        (startPressure, { pressureUnit }) => {
          if (!helpers.req(startPressure)) return true;
          if (typeof startPressure !== 'number') return false;
          return (
            startPressure > 0 &&
            startPressure <= (pressureUnit === PressureUnit.Bar ? 300 : 4400)
          );
        },
      ),
    },
    endPressure: {
      required: helpers.withMessage('End pressure is required', required),
      valid: helpers.withMessage(
        'End pressure must be greater than 0 and cannot be greater than the start pressure',
        (endPressure, { startPressure }) => {
          if (!helpers.req(endPressure)) return true;
          if (typeof endPressure !== 'number') return false;
          return endPressure > 0 && endPressure < startPressure;
        },
      ),
    },
    o2Percentage: {
      between: helpers.withMessage(
        'O₂ percentage must be between 0 and 100',
        between(0, 100),
      ),
      gasMixLimits: helpers.withMessage(
        'O₂ and helium content cannot add to more than 100% of the gas mix',
        (o2Percentage, { hePercentage }) => {
          if (
            typeof o2Percentage === 'number' &&
            typeof hePercentage === 'number'
          ) {
            return o2Percentage + hePercentage <= 100;
          }

          return true;
        },
      ),
    },
    hePercentage: {
      between: helpers.withMessage(
        'Helium percentage must be between 0 and 100',
        between(0, 100),
      ),
    },
  },
  formData,
);

function onTogglePressureUnit() {
  formData.pressureUnit =
    formData.pressureUnit === PressureUnit.Bar
      ? PressureUnit.PSI
      : PressureUnit.Bar;
}

watch(formData, (newData, oldData) => {
  // if (newData.tankId !== oldData.tankId) {
  newData.tankInfo = props.tanks.find((tank) => tank.id === newData.tankId);
  // }
  emit('update', newData, props.ordinal);
});
</script>
