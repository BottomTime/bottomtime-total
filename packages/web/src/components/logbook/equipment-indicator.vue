<template>
  <p class="flex gap-2 items-baseline">
    <label class="font-bold text-right min-w-16">{{ label }}:</label>
    <span v-if="value === true" class="text-success">
      <i class="fas fa-check-circle"></i>
      <span class="sr-only">Yes</span>
    </span>
    <span v-else-if="value === false" class="text-secondary">
      <i class="fas fa-times-circle"></i>
      <span class="sr-only">No</span>
    </span>
    <span v-else class="sr-only">Unspecified</span>
  </p>
</template>

<script lang="ts" setup>
interface EquipmentIndicatorProps {
  label: string;
  value?: boolean;
}

defineProps<EquipmentIndicatorProps>();
</script>
