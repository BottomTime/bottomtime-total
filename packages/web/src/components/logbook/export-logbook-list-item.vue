<template>
  <li class="flex gap-3">
    <FormCheckbox
      :model-value="entry.selected ?? false"
      @update:model-value="$emit('toggle-select', entry, !entry.selected)"
    >
      <span class="sr-only">
        Select entry {{ entry.logNumber }} from
        {{ dayjs(entry.timing.entryTime).format('LLL') }} ({{
          entry.timing.timezone
        }})
      </span>
    </FormCheckbox>

    <div class="grow">
      <div>
        <a
          :id="`e-${entry.id}`"
          :href="`#e-${entry.id}`"
          class="space-x-2 font-title text-lg"
        >
          <span v-if="entry.logNumber" class=""> #{{ entry.logNumber }} </span>
          <span>{{ dayjs(entry.timing.entryTime).format('LLL') }}</span>
        </a>
      </div>

      <div class="mx-2 grid grid-cols-3 gap-3">
        <div v-if="entry.site" class="text-center">
          <p class="font-bold">Dive Site / Location</p>
          <div class="capitalize">
            <p>{{ entry.site.name }},</p>
            <p>{{ entry.site.location }}</p>
          </div>
        </div>

        <div class="text-center">
          <p class="font-bold">Duration</p>
          <p>
            <DurationText :duration="entry.timing.duration" />
          </p>
        </div>

        <div v-if="entry.depths?.maxDepth" class="text-center">
          <p class="font-bold">Max Depth</p>
          <p>
            <DepthText
              :depth="entry.depths.maxDepth"
              :unit="entry.depths.depthUnit"
            />
          </p>
        </div>
      </div>

      <div v-if="entry.tags?.length">
        <FormTags :tags="entry.tags" readonly />
      </div>
    </div>
  </li>
</template>

<script lang="ts" setup>
import { VerySuccinctLogEntryDTO } from '@bottomtime/api';

import { Selectable } from 'src/common';
import dayjs from 'src/dayjs';

import DepthText from '../common/depth-text.vue';
import DurationText from '../common/duration-text.vue';
import FormCheckbox from '../common/form-checkbox.vue';
import FormTags from '../common/form-tags.vue';

interface ExportLogbookListItemProps {
  entry: Selectable<VerySuccinctLogEntryDTO>;
}

defineProps<ExportLogbookListItemProps>();
defineEmits<{
  (
    e: 'toggle-select',
    entry: Selectable<VerySuccinctLogEntryDTO>,
    selected: boolean,
  ): void;
}>();
</script>
