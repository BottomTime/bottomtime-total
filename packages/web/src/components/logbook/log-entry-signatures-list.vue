<template>
  <div>
    <p>
      <span>Showing </span>
      <span class="font-bold">{{ signatures.data.length }}</span>
      <span> of </span>
      <span class="font-bold">{{ signatures.totalCount }}</span>
      <span> signatures.</span>
    </p>
    <ul>
      <li
        v-if="signatures.data.length === 0"
        class="text-lg text-center italic space-x-2 my-3"
      >
        <span>
          <i class="fa-solid fa-circle-exclamation"></i>
        </span>
        <span>Noone has signed this log entry yet.</span>
      </li>

      <li
        v-for="signature in signatures.data"
        :key="signature.id"
        class="flex gap-3 p-2"
      >
        <figure>
          <UserAvatar :profile="signature.buddy" size="small" show-name />
        </figure>

        <div class="grow">
          <p class="font-bold">
            {{ getBuddyTypeString(signature.type) }}
          </p>
          <p
            v-if="signature.agency && signature.certificationNumber"
            class="font-mono text-xs"
          >
            {{ signature.agency.name }} #{{ signature.certificationNumber }}
          </p>
        </div>

        <div>
          {{ dayjs(signature.signedOn).format('LL') }}
        </div>

        <div v-if="editMode">
          <FormButton type="danger" @click="$emit('delete', signature)">
            <span>
              <i class="fa-solid fa-trash"></i>
            </span>
            <span class="sr-only">
              Delete signature from
              {{ signature.buddy.name || `@${signature.buddy.username}` }}
            </span>
          </FormButton>
        </div>
      </li>

      <li v-if="signatures.data.length < signatures.totalCount">
        <LoadingSpinner
          v-if="isLoadingMore"
          message="Fetching more signatures..."
        />
        <a v-else @click="$emit('load-more')">Load more...</a>
      </li>
    </ul>
  </div>
</template>

<script lang="ts" setup>
import { ApiList, BuddyType, LogEntrySignatureDTO } from '@bottomtime/api';

import dayjs from 'src/dayjs';

import FormButton from '../common/form-button.vue';
import LoadingSpinner from '../common/loading-spinner.vue';
import UserAvatar from '../users/user-avatar.vue';

interface LogEntrySignaturesListProps {
  editMode?: boolean;
  isLoadingMore?: boolean;
  signatures: ApiList<LogEntrySignatureDTO>;
}

withDefaults(defineProps<LogEntrySignaturesListProps>(), {
  editMode: false,
  isLoadingMore: false,
});
defineEmits<{
  (e: 'load-more'): void;
  (e: 'delete', signature: LogEntrySignatureDTO): void;
}>();

function getBuddyTypeString(type: BuddyType): string {
  switch (type) {
    case BuddyType.Buddy:
      return 'Buddy';
    case BuddyType.Divemaster:
      return 'Dive Master';
    case BuddyType.Instructor:
      return 'Instructor';
  }
}
</script>
