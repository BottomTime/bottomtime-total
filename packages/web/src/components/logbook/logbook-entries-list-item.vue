<template>
  <li
    class="flex space-x-3 min-h-24 items-center even:bg-blue-300/40 even:dark:bg-blue-900/40 rounded-md p-4"
  >
    <div class="min-w-16 flex">
      <FormCheckbox
        v-if="editMode"
        :control-id="`toggle-${entry.id}`"
        :test-id="`toggle-${entry.id}`"
      />
      <p v-if="entry.logNumber" class="font-bold">#{{ entry.logNumber }}</p>
    </div>

    <div class="flex flex-col gap-3 grow">
      <p
        class="flex flex-col gap-0.5 md:flex-row md:gap-2 items-center md:items-baseline w-full"
      >
        <FormButton
          type="link"
          size="2xl"
          :test-id="`select-${entry.id}`"
          @click="$emit('select', entry)"
        >
          {{ dayjs(entry.timing.entryTime).format('LLL') }}
        </FormButton>
        <span class="italic text-sm">{{ entry.timing.timezone }}</span>
      </p>

      <div
        class="flex flex-col justify-start gap-0 lg:flex-row lg:justify-between"
      >
        <div v-if="entry.timing.bottomTime" class="flex space-x-2">
          <p class="font-bold min-w-40 text-right lg:min-w-0 lg:text-left">
            Bottom time / Duration:
          </p>
          <span class="italic">
            {{ `${entry.timing.bottomTime.toFixed(1)}min` }} /
            {{ `${entry.timing.duration.toFixed(1)}min` }}
          </span>
        </div>

        <div v-else class="flex space-x-2">
          <p class="font-bold min-w-40 text-right lg:min-w-0 lg:text-left">
            Duration:
          </p>
          <span class="italic">
            {{ `${entry.timing.duration.toFixed(1)}min` }}
          </span>
        </div>

        <div v-if="entry.depths?.maxDepth" class="flex space-x-2">
          <p class="font-bold min-w-40 text-right lg:min-w-0 lg:text-left">
            Max depth:
          </p>
          <DepthText
            class="italic"
            :depth="entry.depths.maxDepth"
            :unit="entry.depths.depthUnit || DepthUnit.Meters"
          />
        </div>
      </div>

      <div>
        <p v-if="entry.site" class="flex space-x-2">
          <span class="text-danger">
            <i class="fa-solid fa-location-dot"></i>
          </span>
          <span class="capitalize">
            {{ entry.site.name }}
          </span>
          <span>({{ entry.site.location }})</span>
        </p>
      </div>
    </div>
  </li>
</template>

<script lang="ts" setup>
import { DepthUnit, LogEntryDTO } from '@bottomtime/api';

import dayjs from 'dayjs';

import DepthText from '../common/depth-text.vue';
import FormButton from '../common/form-button.vue';
import FormCheckbox from '../common/form-checkbox.vue';

interface LogbookEntriesListItemProps {
  editMode?: boolean;
  entry: LogEntryDTO;
}

withDefaults(defineProps<LogbookEntriesListItemProps>(), {
  editMode: false,
});
defineEmits<{
  (e: 'select', entry: LogEntryDTO): void;
}>();
</script>
