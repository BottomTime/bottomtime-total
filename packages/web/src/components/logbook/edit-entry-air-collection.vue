<template>
  <ConfirmDialog
    :visible="state.showConfirmRemoveDialog"
    title="Remove Air Tank?"
    icon="fa-regular fa-circle-question fa-2xl"
    confirm-text="Remove"
    @confirm="onConfirmRemoveEntry"
    @cancel="onCancelRemoveEntry"
  >
    <p>Are you sure you want to remove the selected air tank?</p>
  </ConfirmDialog>

  <div class="space-y-2">
    <TransitionGroup name="list" tag="ul" class="space-y-3">
      <EditEntryAir
        v-for="(airEntry, index) in air"
        :key="airEntry.id"
        :air="airEntry"
        :tanks="tanks"
        :ordinal="index"
        @remove="onRemoveEntry"
        @update="(update) => $emit('update', update)"
      />
    </TransitionGroup>

    <div class="flex justify-center space-x-3">
      <FormButton
        control-id="btn-add-tank"
        test-id="btn-add-tank"
        @click="onAddTank"
      >
        <p class="space-x-1">
          <span>
            <i class="fa-solid fa-plus"></i>
          </span>
          <span>Add tank</span>
        </p>
      </FormButton>
    </div>
  </div>
</template>

<script lang="ts" setup>
import { PressureUnit, TankDTO } from '@bottomtime/api';

import { v7 as uuid } from 'uuid';
import { reactive } from 'vue';

import { useCurrentUser } from '../../store';
import FormButton from '../common/form-button.vue';
import ConfirmDialog from '../dialog/confirm-dialog.vue';
import { EditEntryAirFormData } from './edit-entry-air-form-data';
import EditEntryAir from './edit-entry-air.vue';

interface EditEntryAirCollectionProps {
  tanks: TankDTO[];
  air: EditEntryAirFormData[];
}

interface EditEntryAirCollectionState {
  showConfirmRemoveDialog: boolean;
  entryToRemove?: EditEntryAirFormData;
}

const BlankAirForm: EditEntryAirFormData = {
  id: '',
  startPressure: '',
  endPressure: '',
  count: '',
  pressureUnit: PressureUnit.Bar,
  hePercent: '',
  o2Percent: '',
  tankId: '',
} as const;

const currentUser = useCurrentUser();

const props = defineProps<EditEntryAirCollectionProps>();
const emit = defineEmits<{
  (e: 'add', air: EditEntryAirFormData): void;
  (e: 'remove', id: string): void;
  (e: 'update', air: EditEntryAirFormData): void;
}>();
const state = reactive<EditEntryAirCollectionState>({
  showConfirmRemoveDialog: false,
});

function onAddTank() {
  emit('add', {
    ...BlankAirForm,
    id: uuid(),
    pressureUnit: currentUser.user?.settings.pressureUnit ?? PressureUnit.Bar,
  });
}

function onRemoveEntry(id: string) {
  state.entryToRemove = props.air.find((entry) => entry.id === id);
  if (state.entryToRemove) state.showConfirmRemoveDialog = true;
}

function onCancelRemoveEntry() {
  state.entryToRemove = undefined;
  state.showConfirmRemoveDialog = false;
}

function onConfirmRemoveEntry() {
  if (state.entryToRemove) {
    emit('remove', state.entryToRemove.id);
  }
  state.showConfirmRemoveDialog = false;
}
</script>

<style scoped>
.list-enter-active,
.list-leave-active {
  transition: all 0.5s ease;
}
.list-enter-from,
.list-leave-to {
  opacity: 0;
  transform: translateX(30px);
}
</style>
