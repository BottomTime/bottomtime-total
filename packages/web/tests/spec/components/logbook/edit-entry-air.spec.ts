import {
  ListTanksResponseDTO,
  ListTanksResponseSchema,
  PressureUnit,
} from '@bottomtime/api';

import { ComponentMountingOptions, mount } from '@vue/test-utils';

import { EditEntryAirFormData } from '../../../../src/components/logbook/edit-entry-air-form-data';
import EditEntryAir from '../../../../src/components/logbook/edit-entry-air.vue';
import TankData from '../../../fixtures/tanks.json';

const TestId = '42eee38a-5579-4e9d-b738-7126b7682b97';
const RemoveButton = '[data-testid="remove-tank"]';
const TankSelect = `#tanks-select-${TestId}`;
const Doubles = `#doubles-${TestId}`;
const TankSummary = '[data-testid="tank-summary"]';
const StartPressure = `#start-pressure-${TestId}`;
const EndPressure = `#end-pressure-${TestId}`;
const OxygenPercentage = `#o2-${TestId}`;
const HeliumPercentage = `#he-${TestId}`;

describe('EditEntryAir component', () => {
  let opts: ComponentMountingOptions<typeof EditEntryAir>;
  let tankData: ListTanksResponseDTO;
  let air: EditEntryAirFormData;

  beforeAll(() => {
    tankData = ListTanksResponseSchema.parse(TankData);
  });

  beforeEach(() => {
    air = {
      count: 1,
      endPressure: 800,
      hePercent: 12,
      id: TestId,
      o2Percent: 32,
      pressureUnit: PressureUnit.PSI,
      startPressure: 3000,
      tankId: tankData.tanks[0].id,
      tankInfo: {
        ...tankData.tanks[0],
      },
    };

    opts = {
      props: {
        tanks: tankData.tanks,
        ordinal: 2,
        air,
      },
    };
  });

  it('will render correctly', () => {
    const wrapper = mount(EditEntryAir, opts);
    expect(wrapper.get<HTMLSelectElement>(TankSelect).element.value).toBe(
      air.tankId,
    );
    expect(wrapper.get<HTMLInputElement>(Doubles).element.checked).toBe(false);
    expect(wrapper.get(TankSummary).html()).toMatchSnapshot();
    expect(wrapper.get<HTMLInputElement>(StartPressure).element.value).toBe(
      air.startPressure.toString(),
    );
    expect(wrapper.get<HTMLInputElement>(EndPressure).element.value).toBe(
      air.endPressure.toString(),
    );
    expect(wrapper.get<HTMLInputElement>(OxygenPercentage).element.value).toBe(
      air.o2Percent.toString(),
    );
    expect(wrapper.get<HTMLInputElement>(HeliumPercentage).element.value).toBe(
      air.hePercent.toString(),
    );
  });

  it('will render a blank air entry correctly', () => {
    const blank: EditEntryAirFormData = {
      count: '',
      endPressure: '',
      hePercent: '',
      id: TestId,
      o2Percent: '',
      pressureUnit: PressureUnit.Bar,
      startPressure: '',
      tankId: '',
    };
    opts.props = {
      tanks: tankData.tanks,
      ordinal: 2,
      air: blank,
    };
    const wrapper = mount(EditEntryAir, opts);

    expect(wrapper.get<HTMLSelectElement>(TankSelect).element.value).toBe('');
    expect(wrapper.get<HTMLInputElement>(Doubles).element.checked).toBe(false);
    expect(wrapper.find(TankSummary).exists()).toBe(false);
    expect(wrapper.get<HTMLInputElement>(StartPressure).element.value).toBe('');
    expect(wrapper.get<HTMLInputElement>(EndPressure).element.value).toBe('');
    expect(wrapper.get<HTMLInputElement>(OxygenPercentage).element.value).toBe(
      '',
    );
    expect(wrapper.get<HTMLInputElement>(HeliumPercentage).element.value).toBe(
      '',
    );
  });

  it('will render correctly if saved tank info does not match an item in the list', () => {
    opts.props = {
      tanks: tankData.tanks,
      ordinal: 2,
      air: {
        ...air,
        tankId: '',
      },
    };
    const wrapper = mount(EditEntryAir, opts);
    expect(wrapper.get<HTMLSelectElement>(TankSelect).element.value).toBe(
      'current',
    );
    expect(wrapper.get<HTMLInputElement>(Doubles).element.checked).toBe(false);
    expect(wrapper.get(TankSummary).html()).toMatchSnapshot();
    expect(wrapper.get<HTMLInputElement>(StartPressure).element.value).toBe(
      air.startPressure.toString(),
    );
    expect(wrapper.get<HTMLInputElement>(EndPressure).element.value).toBe(
      air.endPressure.toString(),
    );
    expect(wrapper.get<HTMLInputElement>(OxygenPercentage).element.value).toBe(
      air.o2Percent.toString(),
    );
    expect(wrapper.get<HTMLInputElement>(HeliumPercentage).element.value).toBe(
      air.hePercent.toString(),
    );
  });

  it('will render with "Doubles" checked if count is 2', () => {
    opts.props = {
      tanks: tankData.tanks,
      ordinal: 2,
      air: {
        ...air,
        count: 2,
      },
    };
    const wrapper = mount(EditEntryAir, opts);
    expect(wrapper.get<HTMLInputElement>(Doubles).element.checked).toBe(true);
  });

  it('will emit remove event if remove button is clicked', async () => {
    const wrapper = mount(EditEntryAir, opts);
    await wrapper.get(RemoveButton).trigger('click');
    expect(wrapper.emitted('remove')).toEqual([[air.id]]);
  });
});
