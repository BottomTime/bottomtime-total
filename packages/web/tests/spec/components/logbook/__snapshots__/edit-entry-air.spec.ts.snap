// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditEntryAir component will render correctly 1`] = `
"<div class="flex justify-evenly col-span-1 md:col-span-2 lg:col-span-4 order-3 md:order-3 mb-2" data-testid="tank-summary">
  <div class="text-center">
    <p class="font-bold">Working Pressure</p>
    <p class="font-mono text-sm">207bar </p>
  </div>
  <div class="text-center">
    <p class="font-bold">Volume</p>
    <p class="font-mono text-sm">1.8L</p>
  </div>
  <div class="text-center">
    <p class="font-bold">Material</p>
    <p class="font-mono text-sm">Aluminum</p>
  </div>
</div>"
`;

exports[`EditEntryAir component will render correctly if saved tank info does not match an item in the list 1`] = `
"<div class="flex justify-evenly col-span-1 md:col-span-2 lg:col-span-4 order-3 md:order-3 mb-2" data-testid="tank-summary">
  <div class="text-center">
    <p class="font-bold">Working Pressure</p>
    <p class="font-mono text-sm">207bar </p>
  </div>
  <div class="text-center">
    <p class="font-bold">Volume</p>
    <p class="font-mono text-sm">1.8L</p>
  </div>
  <div class="text-center">
    <p class="font-bold">Material</p>
    <p class="font-mono text-sm">Aluminum</p>
  </div>
</div>"
`;
